src/ripple/app/misc/AMMHelpers.h

```diff
@@ -0,0 +1,224 @@
+#include <ripple/protocol/Issue.h>
+#include <ripple/protocol/Quality.h>
+#include <ripple/protocol/STAccount.h>
+#include <ripple/protocol/STAmount.h>
+
+#include <type_traits>
+
+namespace ripple {
+
+/** Calculate LP Tokens given AMM pool reserves.
+ * @param asset1 AMM one side of the pool reserve
+ * @param asset2 AMM another side of the pool reserve
+ * @return LP Tokens as IOU
+ */
+STAmount
+ammLPTokens(
+    STAmount const& asset1,
+    STAmount const& asset2,
+    Issue const& lptIssue);
+
+/** Convert to the fee from the basis points
+ * @param tfee  trading fee in {0, 1000}
+ * 1 = 1/10bps or 0.001%, 1000 = 1%
+ */
+inline Number
+getFee(std::uint16_t tfee)
+{
+    return Number{tfee} / 100000;
+}
+
+/** Get fee multiplier (1 - tfee)
+ * @tfee trading fee in basis points
+ */
+inline Number
+feeMult(std::uint16_t tfee)
+{
+    return 1 - getFee(tfee);
+}
+
+/** Get fee multiplier (1 - tfee / 2)
+ * @tfee trading fee in basis points
+ */
+inline Number
+feeMultHalf(std::uint16_t tfee)
+{
+    return 1 - getFee(tfee) / 2;
+}
+
+/** Calculate LP Tokens given asset's deposit amount.
+ * @param asset1Balance current AMM asset1 balance
+ * @param asset1Deposit requested asset1 deposit amount
+ * @param lpTokensBalance LP Tokens balance
+ * @param tfee trading fee in basis points
+ * @return tokens
+ */
+STAmount
+lpTokensIn(
+    STAmount const& asset1Balance,
+    STAmount const& asset1Deposit,
+    STAmount const& lpTokensBalance,
+    std::uint16_t tfee);
+
+/** Calculate asset deposit given LP Tokens.
+ * @param asset1Balance current AMM asset1 balance
+ * @param lpTokensBalance LP Tokens balance
+ * @param ammTokensBalance AMM LPT balance
+ * @param tfee trading fee in basis points
+ * @return
+ */
+STAmount
+ammAssetIn(
+    STAmount const& asset1Balance,
+    STAmount const& lpTokensBalance,
+    STAmount const& ammTokensBalance,
+    std::uint16_t tfee);
+
+/** Calculate LP Tokens given asset's withdraw amount. Return 0
+ * if can't calculate.
+ * @param asset1Balance current AMM asset1 balance
+ * @param asset1Withdraw requested asset1 withdraw amount
+ * @param lpTokensBalance LP Tokens balance
+ * @param tfee trading fee in basis points
+ * @return tokens out amount
+ */
+STAmount
+lpTokensOut(
+    STAmount const& asset1Balance,
+    STAmount const& asset1Withdraw,
+    STAmount const& lpTokensBalance,
+    std::uint16_t tfee);
+
+/** Calculate asset withdrawal by tokens
+ * @param assetBalance balance of the asset being withdrawn
+ * @param lptAMMBalance total AMM Tokens balance
+ * @param lpTokens LP Tokens balance
+ * @param tfee trading fee in basis points
+ * @return calculated asset amount
+ */
+STAmount
+withdrawByTokens(
+    STAmount const& assetBalance,
+    STAmount const& lptAMMBalance,
+    STAmount const& lpTokens,
+    std::uint32_t tfee);
+
+/** Find in/out amounts to change the spot price quality to the requested
+ * quality. Implements AMM Swap equation (11) and calls swapAssetIn
```

Comment:
I'm not sure "Implements AMM Swap equation (11)" is a useful comment unless you link to the document this refers to, and the document should probably live in this repo. If we do that, we should probably give names to the equations rather than numbers

Reply:
This is my fault, because I asked him to add the equation numbers from the spec, but didn't think about linking to the spec. 😀 FWIW, I think a URL would be sufficient. e.g "Implements AMM Swap equation (11) from [https://github.com/XRPLF/XRPL-Standards/discussions/78], and calls [....]"

Reply:
I changed to `Implements AMM change swap price, which returns takerPays.
Then swaps in to find takerGets.`. There are quite a few places with AMM equations, should the URL be added everywhere the equations are used?

Reply:
I think it's reasonable to include the URL everywhere.